### DISCUSSION TITLE
Does SolidJS have an official store like Vuex/Pinia and Redux?

### DISCUSSION THREAD

**Discussion by yandodov** [solidjs/solid#1595] [Q&A]
I'm new to SolidJS and I'm thinking about using it. However, I'm not sure how I'd handle shared state. Is there an equivalent to Vuex and Redux, but for Solid? Or it's too cool for that and you can somehow just use the reactive primitives it provides?

**Comment by atk** [comment]
Signals and stores are pretty powerful and don't require as much boilerplate as redux, nor do they encourage unnecessary complexity like rxjs.

Usually, Solid's state primitives should enable you to handle your use cases, but if not, observable and from allow binding other state management libraries into your apps.

**Reply by edimoldovan** [reply]
I have some similar questions, so I'll add them here since they relate to store/state management:
- so we have createSignal and createReource -- where can I find examples of an actual example of how to handle real data flow in a Solid app: UI triggers a change, say a function somewhere reacts to that change, changes the state (myriad of questions here aroun where is that state), sends the state change for persistence to an API
- then, on reload, it does the opposite, reads the state from the api, rebuilds the local state and then from that updates the UI

I have been looking and trying out various approaches with createResource and createSignal, route data, but also createStore -- the amount of reactivity options is just confusing and none of the examples I found show anything end to end (been trying to find examples for days now).

So, how should I be approaching this? I am not necessarily looking for code, more of an architecture I can follow and implement. Happy to also share if I made it happen.

**Reply by atk** [reply]
> How to handle real data flow in a Solid app:

First of all, you should organize your state in the simplest possible manner. If a simple bool/number/string will do, use createSignal, if your data actually needs more complexity, use createStore and if you need asynchronous data, createReasource is the way to go.

Next, in some cases, you need to derive data. You should only use createMemo if you want to stop effects from running even though the subscribed state changed. Otherwise, a simple function will do.

Finally, the simplest way to persist your data in localStorage is to wrap your setter in a function that will put the current version into it:

```js
function createPersistence(key) { return [
  function getInitialValue() {
    const initialValueString = localStorage.get(key);
    return initialValueString === null ? undefined : JSON.parse(initialValueString);
  },
  function setPersistedValue(setter) {
    return function(...setterArgs) {
      localStorage.set(key, setter(...setterArgs));
    };
  }
]};

// use it like this (same with a store):
const [initialSignalValue, setPersistedValue] = createPersistence('signal');
const [persistedSignal, _setPersistedSignal] = createSignal(initialSignalValue());
const setPersistedSignal = setPersistedValue(_setPersistedSignal);
```

Now for triggering changes: usually, events or other side effects will change the state. In solid, the state changes propagate automatically through all effects. The parts of JSX between `{}` are automatically wrapped in effects, too.
So the state changes and directly calls the effects that registered their usage of the state by using it when the component first ran.
